process: make mypy happy with type annotations

plamut opened the pull request:
Closes #987.

**PR checklist:**
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

plamut committed:
process: add mypy types check to nox sessions

plamut committed:
Ignore type errors for not annotated modules

Several dependencies lack type annotations, or they don't advertise
themselves as type-annotated. We do not want `mypy` to complain about
these.

plamut committed:
Fix mypy complaints (batch 1)

plamut committed:
Fix mypy complaints (batch 2)

plamut committed:
Fix mypy complaints (batch 3)

plamut committed:
Fix mypy false positive errors

tswast commented on noxfile.py:
```diff
@@ -108,9 +109,24 @@ def unit_noextras(session):
 
 
+@nox.session(python=DEFAULT_PYTHON_VERSION)
+def mypy(session):
```
Can you also add `mypy` to `nox.options.sessions` so that it runs in Kokoro?

plamut replied:
Ah, I thought it would be picked automatically/ Sure thing, will add. :+1:

plamut committed:
Merge branch 'main' into iss-987

plamut committed:
Simplify external config options instantiation

plamut committed:
Do not ignore api-core in type checks

More recent releases of google-api-core have typing enabled.

plamut commented on google/cloud/bigquery/job/base.py:
```diff
@@ -21,8 +21,8 @@
 
-from google.api_core import exceptions
-import google.api_core.future.polling
+from google.api_core import exceptions  # type: ignore
```
**Update:** Works well, meaning that related `type: ignore` comments can be removed.

plamut committed:
Remove unneeded __hash__ = None lines

plamut committed:
Use an alias for timeout type in client.py

plamut committed:
Fix PathLike subscription error in pre-Python 3.9

plamut commented on google/cloud/bigquery/client.py:
```diff
@@ -526,12 +538,12 @@ def _ensure_bqstorage_client(
                 credentials=self._credentials,
                 client_options=client_options,
-                client_info=client_info,
+                client_info=client_info,  # type: ignore  # (None is also accepted)
```
It appears so, yes. The [docstring](https://github.com/googleapis/python-bigquery-storage/blob/acc92249013f1b31fdac2aa4bf5a6864730d7422/google/cloud/bigquery_storage_v1/services/big_query_read/client.py#L305-L309) says that `None` is allowed, but `mypy` complains that `Optional[...ClientInfo]` is not acceptable here as an argument.

It's probably because the default value of `client_info` is not a literal `None`, but `DEFAULT_CLIENT_INFO` instead.

plamut commented on google/cloud/bigquery/client.py:
```diff
@@ -1934,7 +1944,7 @@ def create_job(
                 source_uris,
                 destination,
-                job_config=load_job_config,
+                job_config=typing.cast(LoadJobConfig, load_job_config),
```
The problem is that `from_api_repr()` is defined on the base `_JobConfig` class and its return type is annotated as `_JobConfig`. Which makes sense on the base class, but in practice the actual `cls` subclass is chosen here dynamically by introspecting `job_config`, and `mypy` cannot understand that.

Specifying the type of `load_job_config` at assignment time would not help, because again a base type (`_JobConfig`) is assigned to `load_job_config`  with a specific sub-type, and `mypy` would still complain. Cast solves this.

The alternative could be changing the return type on the base class to include all sub-classes, but that doesn't feel right. It can get of of sync if a new sub-type is created, and a circular import would be needed.
Besides, `mypy` would probably _still_ complain if `"Union[All, Config, Subtypes]"` is assigned to a _specific_ sub-class type.

plamut commented on google/cloud/bigquery/client.py:
```diff
@@ -2064,15 +2077,19 @@ def get_job(
         )
 
-        return self.job_from_resource(resource)
+        job_instance = self.job_from_resource(resource)  # never an UnknownJob
```
Hm... I was under impression that `job_from_resource()` is a helper that only returns `UnknownJob` if somebody manually provides an "invalid"/incomplete resource dict.

However, if the backend can also return such resource (e.g. due to permissions), the assumption in the comment does not hold (although the return type in the docstring already excludes `UnknownJob`).

Do you maybe remember the combination of permissions that result in an `UnknownJob`?
(also, I might need assistance here, because my access to cloud console is now severely limited - I'll have to check)

tswast replied:
> Do you maybe remember the combination of permissions that result in an UnknownJob?
(also, I might need assistance here, because my access to cloud console is now severely limited - I'll have to check)

It related to creating a job with one account and trying to list all jobs for the project with another account.

See: `jobs.list` versus `jobs.listAlll` here: https://cloud.google.com/bigquery/docs/access-control




Permission | Description
-- | --
bigquery.jobs.list | List all jobs and retrieve metadata on any job submitted by any user.1 For jobs submitted by other users, details and metadata are redacted.
bigquery.jobs.listAll | List all jobs and retrieve metadata on any job submitted by any user.1

1 For any job you create, you automatically have the equivalent of the bigquery.jobs.get and bigquery.jobs.update permissions for that job.

plamut replied:
@tswast My access to Cloud Console is limited now and I cannot create an account anymore that only has `bigquery.jobs.list` permission (and possibly a few others to access jobs on the project), while the current development account that I use has the `bigquery.jobs.listAll` permission.

In other words, I don't think I can actually retrieve an `UnknownJob`.

Could you run the below script and see what happens if we call `get_job()` on an `UnknownJob`?

<details>
<summary>unknown_jobs.py</summary>

```py
from collections import Counter

from google.cloud import bigquery
from google.oauth2 import service_account


PROJECT_ID = "TODO: set"
DATASET_ID = "TODO: set"
TABLE_ID = "TODO: set"

MAIN_CRED_FILE = "TODO: /path/to/service_account.json"

# NOTE: This one must have "bigquery.jobs.list" permission, but NOT "bigquery.jobs.listAll".
# We want metadata of jobs submitted by other users redacted.
OTHER_CRED_FILE = "TODO: /path/to/other/service_account.json"


def main():
    credentials = service_account.Credentials.from_service_account_file(MAIN_CRED_FILE)
    client = bigquery.Client(credentials=credentials)

    # Create table if needed.
    table_name_full = f"{PROJECT_ID}.{DATASET_ID}.{TABLE_ID}"
    table = client.create_table(table_name_full, exists_ok=True)

    # Create at least one job (load job in this case).
    job_config = bigquery.LoadJobConfig(
        schema=[
            bigquery.SchemaField("user_id", "STRING"),
            bigquery.SchemaField("name", "STRING"),
        ],
        write_disposition=bigquery.WriteDisposition.WRITE_TRUNCATE,
    )
    row = {"user_id": "1234567890", "name": "John"}
    load_job = client.load_table_from_json([row], destination=table, job_config=job_config)
    result = load_job.result()
    print(result)

    print("=" * 20)

    # Using another client, use other users' jobs. The client should NOT have
    # "bigquery.jobs.listAll" permission.
    credentials = service_account.Credentials.from_service_account_file(OTHER_CRED_FILE)
    other_client = bigquery.Client(credentials=credentials)

    jobs_iter = other_client.list_jobs(
        project=PROJECT_ID, max_results=10_000, all_users=True
    )
    retrieved_jobs = list(jobs_iter)

    jobs_count = Counter(type(job) for job in retrieved_jobs)
    print(jobs_count)

    try:
        unknown_job = next(
            job for job in retrieved_jobs if isinstance(job, bigquery.UnknownJob)
        )
    except StopIteration:
        print("No UnknownJobs received, cannot test get_job().")
        return
    
    unknown_job = other_client.get_job(unknown_job)
    print(unknown_job._properties)


if __name__ == "__main__":
    main()
```
</details>

plamut replied:
FWIW, I preemptively changed the return type annotation to also include `UnknownJob`.

plamut committed:
Fix a typo in docstring

Co-authored-by: Tim Swast <swast@google.com>

plamut committed:
Add mypy to the list of nox sessions to run

plamut commented:
A subtlety here is that type checks will only pass of `google-api-core>=2.1.1` is installed when [type check support](https://github.com/googleapis/python-api-core/blob/main/CHANGELOG.md#211-2021-10-13) was added.

As the time passes, this will eventually become a non-issue, but in the transitional period people might experience failures if they have less recent `python-api-core` installed. Those users would likely have to ignore `google-cloud-bigquery` when type-checking their projects.

plamut committed:
Fix opentelemetry type error

The Opentelemetry APi has changed from the minimum version the
BigQuery client currently uses, we thus need to bound the maximum
Opentelemetry version.

In addition, that maximum version does not yet support type checks, thus
it is ignored.

plamut commented on setup.py:
```diff
@@ -63,9 +63,11 @@
-        "opentelemetry-api >= 0.11b0",
-        "opentelemetry-sdk >= 0.11b0",
-        "opentelemetry-instrumentation >= 0.11b0",
+        # The Opentelemetry API was changed in v0.15b, we need to bound the version.
+        # https://github.com/open-telemetry/opentelemetry-python/commit/f3cdfa2cdb9cb1c442189e2ead3788f45d92352d#diff-ebe3b7e2373ce704ff49dcab7a82d2270aa35c624337325cd0233e60e7a95a39R35
+        "opentelemetry-api >= 0.11b0, <0.15b0",
+        "opentelemetry-sdk >= 0.11b0, <0.15b0",
+        "opentelemetry-instrumentation >= 0.11b0, <0.15b0",
```
I discovered this when investigating the alternating `mypy` error and unit test error under Python 3.6 (fixing one produced the other). It turned out that the test/check failure depends on which `opentelemetry-api` version is installed, because the latter's surface has changed since `0.11b` that we pin as a minimum.

We should probably bump the opentelemetry version in a separate PR, though. We would like to use the more up-to-date API, should preferably also be a version that is type-checked (at least [1.1.0](https://github.com/open-telemetry/opentelemetry-python/blob/main/CHANGELOG.md#110---2021-04-20)).

tswast replied:
+1 on bumping opentelemetry in a separate PR.

Note: we might want to consider #1038 to be a bug, since I think it means we can't have Spanner and BigQuery clients installed in the same environment.

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

plamut commented:
It appears some more adjustments are needed to opentelemetry tests (or code).

plamut committed:
Exclude type-checking code from coverage

plamut committed:
Fix patching opentelemetry tracer pvoider

plamut committed:
Merge branch 'main' into iss-987

plamut committed:
Adjust get_job() return type, ignore opentelemetry

plamut committed:
Merge branch 'main' into iss-987

plamut committed:
Merge branch 'main' into iss-987

tswast committed:
Merge branch 'main' into iss-987

tswast commented:
🎉 

```
nox > Running session mypy
nox > Creating virtual environment (virtualenv) using python3.8 in .nox/mypy
nox > python -m pip install -e .[all]
nox > python -m pip install ipython
nox > python -m pip install mypy==0.910
nox > python -m pip install types-protobuf types-python-dateutil types-requests types-setuptools
nox > mypy google/cloud
Success: no issues found in 49 source files
nox > Session mypy was successful.
```
